core/src/main/java/org/infinispan/CacheStream.java

```diff
@@ -0,0 +1,251 @@
+    * is not specified.  Please see {@link CacheStream#iterator()} for more information.</p>
+    * <p>Multiple listeners may be registered upon multiple invocations of this method.  The ordering of notified
+    * listeners is not specified.</p>
+    * @param listener The listener that will be called back as segments are completed.
+    * @return This stream again with the listener registered.
+    */
+   CacheStream<R> segmentCompletionListener(SegmentCompletionListener listener);
+
+   /**
+    * Disables tracking of rehash events that could occur to the underlying cache.  If a rehash event occurs while
+    * a terminal operation is being performed it is possible for some values that are in the cache to not be found.
+    * Note that you will never have an entry duplicated when rehash awareness is disabled, only lost values.
+    * <p>Most terminal operations will run faster with rehash awareness disabled even without a rehash occuring.
+    * However if a rehash occurs with this disabled be prepared to possibly receive only a subset of values.</p>
+    * @return This stream again with rehash awareness disabled.
+    */
+   CacheStream<R> disableRehashAware();
+
+   /**
+    * Functional interface that is used as a callback when segments are completed.  Please see
+    * {@link CacheStream#segmentCompletionListener(SegmentCompletionListener)} for more details.
+    * @since 8.0
+    */
+   @FunctionalInterface
+   interface SegmentCompletionListener {
+      /**
+       * Method invoked when the segment has been found to be consumed properly by the terminal operation.
+       * @param segment The segments that were completed
+       */
+      void segmentCompleted(Set<Integer> segment);
+   }
+
+   /**
+    * {@inheritDoc}
+    * <p>This operation is performed remotely on the node that is the primary owner for the key tied to the entry(s)
+    * in this stream.</p>
+    * <p>NOTE: This method while being rehash aware has the lowest consistency of all of the operators.  This
+    * operation will be performed on every entry at least once in the cluster, as long as the originator doesn't go
+    * down while it is being performed.  This is due to how the distributed action is performed.  Essentially the
+    * {@link CacheStream#distributedBatchSize} value controls how many elements are processed per node at a time
+    * when rehash is enabled. After those are complete the keys are sent to the originator to confirm that those were
+    * processed.  If that node goes down during/before the response those keys will be processed a second time.</p>
+    * <p>This method is ran distributed by default with a distributed backing cache.  However if you wish for this
+    * operation to run locally you can use the {@link CacheStream#iterator()} method to return all of the results
+    * locally and then use {@link Iterator#forEachRemaining(Consumer)} method for a single threaded variant.  If you
```

Comment:
The explication is kind of long with all those links, I think a single `{@code stream().iterator().forEachRemaining()}` would be easier to parse. Same for the parallel iteration (and maybe include the try-with-resources statement).